pkg/cortex/cortex.go

```diff
@@ -238,50 +247,101 @@ func (t *Cortex) setupAuthMiddleware(cfg *Config) {
 	}
 }
 
-func (t *Cortex) init(cfg *Config, m moduleName) error {
+func (t *Cortex) initModuleServices(cfg *Config, target moduleName) (map[moduleName]services.Service, error) {
+	servicesMap := map[moduleName]services.Service{}
+
 	// initialize all of our dependencies first
-	for _, dep := range orderedDeps(m) {
-		if err := t.initModule(cfg, dep); err != nil {
-			return err
+	deps := orderedDeps(target)
+	deps = append(deps, target) // lastly, initialize the requested module
+
+	for ix, n := range deps {
+		mod := modules[n]
+
+		var serv services.Service
+
+		if mod.service != nil {
+			s, err := mod.service(t, cfg)
+			if err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("error initialising module: %s", n))
+			}
+			serv = s
+		} else if mod.wrappedService != nil {
+			s, err := mod.wrappedService(t, cfg)
+			if err != nil {
+				return nil, errors.Wrap(err, fmt.Sprintf("error initialising module: %s", n))
+			}
+			if s != nil {
+				// We pass servicesMap, which isn't yet finished. By the time service starts,
+				// it will be fully built, so there is no need for extra synchronization.
+				serv = newModuleServiceWrapper(servicesMap, n, s, mod.deps, findReverseDeps(n, deps[ix+1:]))
+			}
 		}
-	}
-	// lastly, initialize the requested module
-	return t.initModule(cfg, m)
-}
 
-func (t *Cortex) initModule(cfg *Config, m moduleName) error {
-	level.Info(util.Logger).Log("msg", "initialising", "module", m)
-	if modules[m].init != nil {
-		if err := modules[m].init(t, cfg); err != nil {
-			return errors.Wrap(err, fmt.Sprintf("error initialising module: %s", m))
+		if serv != nil {
+			servicesMap[n] = serv
 		}
 	}
-	return nil
+
+	return servicesMap, nil
 }
 
 // Run starts Cortex running, and blocks until a signal is received.
 func (t *Cortex) Run() error {
-	return t.server.Run()
-}
+	// get all services, create service manager and tell it to start
+	servs := []services.Service(nil)
+	for _, s := range t.serviceMap {
+		servs = append(servs, s)
+	}
 
-// Stop gracefully stops a Cortex.
-func (t *Cortex) Stop() error {
-	t.stopModule(t.target)
-	deps := orderedDeps(t.target)
-	// iterate over our deps in reverse order and call stopModule
-	for i := len(deps) - 1; i >= 0; i-- {
-		t.stopModule(deps[i])
+	sm, err := services.NewManager(servs...)
+	if err != nil {
+		return err
 	}
-	return nil
-}
 
-func (t *Cortex) stopModule(m moduleName) {
-	level.Info(util.Logger).Log("msg", "stopping", "module", m)
-	if modules[m].stop != nil {
-		if err := modules[m].stop(t); err != nil {
-			level.Error(util.Logger).Log("msg", "error stopping", "module", m, "err", err)
+	// Let's listen for events from this manager, and log them.
+	healthy := func() { level.Info(util.Logger).Log("msg", "Cortex started") }
+	stopped := func() { level.Info(util.Logger).Log("msg", "Cortex stopped") }
+	serviceFailed := func(service services.Service) {
+		// if any service fails, stop entire Cortex
+		sm.StopAsync()
```

Comment:
Shouldn't also wait until stopped?

Reply:
We do, but not here. This is called by service manager, when it detects that any service has failed. In such case, we tell service manager to stop entire service manager (all Cortex services).

Cortex waits until service manager is stopped later in this method:

```
	// this can only return error if context is Done... ignore
	_ = sm.AwaitStopped(context.Background())
```